### PR TITLE
feat(compiler): moves `autoDefineCustomElements` to an export behavior

### PR DESCRIPTION
**Pull request by tanner-reits**
This commit moves `autoDefineCustomElements` from a config flag to a `customElementsExportBehavior` option on the `dist-custom-elements` output target. This prevents treeshaking issues that were possible when barrel exporting and this option were both enabled.

Prep work for STENCIL-457

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [X] Build (`npm run build`) was run locally and any changes were pushed
- [X] Unit tests (`npm test`) were run locally and passed
- [X] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [X] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [X] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Currently, `autoDefineCustomElements` can be enabled while the `single-export-module` export behavior option is set which can break tree-shaking behavior.

GitHub Issue Number: N/A

## What is the new behavior?

This prevents both barrel export behavior and auto-define behavior from co-existing so tree-shaking will always work as intended.

## Does this introduce a breaking change?

- [X] Yes
- [ ] No

Released with Stencil v3

## Testing

Unit and e2e tests continue to pass. Verified function is called in output JS files.

### CODE REVIEW

**Comment by tanner-reits** [comment]
This is some initial work before we move into STENCIL-457

- I'll open up a PR to update the Stencil site docs for V3
- I'll also have a follow-up PR for updating the "breaking changes" docs
- I'll update our internal Notion docs to remove the flag from the config table

**Review by alicewriteswrongs — APPROVED** [review]
looks straightforward enough!

**Review by rwaskiewicz — APPROVED** [review]

File: src/declarations/stencil-public-compiler.ts [hunk 1]
```diff
@@ -2028,8 +2028,15 @@ export interface OutputTargetBaseNext {
...
  * - `single-export-module`: All components will be re-exported from the specified directory's root `index.js` file.
+ * - `auto-define-custom-elements`: Enables the auto-definition of a component and its children (recursively) in the custom elements registry. This
+ * functionality allows consumers to bypass the explicit call to define a component, its children, its children's
+ * children, etc. Users of this flag should be aware that enabling this functionality may increase bundle size.
  */
-export const CustomElementsExportBehaviorOptions = ['default', 'single-export-module'] as const;
+export const CustomElementsExportBehaviorOptions = [
+  'default',
+  'single-export-module',
+  'auto-define-custom-elements',
```
Comment (rwaskiewicz) [inline src/declarations/stencil-public-compiler.ts:2038]: I'm gonna hold off on merging until we answer https://github.com/ionic-team/stencil-site/pull/916#discussion_r974243770, in the event we do wanna change the doc order _and_ if we wanna propagate those changes here

**Comment by rwaskiewicz** [comment]
I think `Tech Debt Burndown / Download error files and report (pull_request_target)` is failing due to `v3.0.0-dev` being a week out of date with `main`, which has led to a large payload that we'd be sending to the comment bot. Once we resolve https://github.com/ionic-team/stencil/pull/3615/files#r974244751 I'll update `v3.0.0-dev` and update the branch against it to see if that fixes it 🔧

**Review by rwaskiewicz — APPROVED** [review]

**Comment by rwaskiewicz** [comment]
https://github.com/ionic-team/stencil/pull/3620 to fixup the v3.0.0 branch (and is a precursor to updating this branch)

**Pull request merged**